salt/utils/dependencies.py

```diff
@@ -0,0 +1,144 @@
+        self.version_getter = version_getter
+        self.has_depend = has_depend
+        self.version = version
+        self.populate()
+
+    @property
+    def module(self):
+        return self.module_getter(self.name)
+
+    def populate(self):
+        if self.has_depend is None:
+            mod = self.module_getter(self.name)
+            if mod:
+                self.has_depend = True
+                self.version = self.version_getter(mod)
+            else:
+                self.has_depend = False
+
+    def __nonzero__(self):
+        return self.has_depend
+
+    def __eq__(self, other):
+        other_ver = LooseVersion(other)
+        dep_ver = LooseVersion(self.version)
+        return dep_ver == other_ver
+
+    def __ne__(self, other):
+        other_ver = LooseVersion(other)
+        dep_ver = LooseVersion(self.version)
+        return dep_ver != other_ver
+
+    def __lt__(self, other):
+        other_ver = LooseVersion(other)
+        dep_ver = LooseVersion(self.version)
+        return dep_ver < other_ver
+
+    def __le__(self, other):
+        other_ver = LooseVersion(other)
+        dep_ver = LooseVersion(self.version)
+        return dep_ver <= other_ver
+
+    def __gt__(self, other):
+        other_ver = LooseVersion(other)
+        dep_ver = LooseVersion(self.version)
+        return dep_ver > other_ver
+
+    def __ge__(self, other):
+        other_ver = LooseVersion(other)
+        dep_ver = LooseVersion(self.version)
+        return dep_ver >= other_ver
+
+
+def msgpack_module_getter(name):
+    msgpack = None
+    try:
+        # Attempt to import msgpack
+        import msgpack
+        if msgpack.version >= (0, 4, 0):
+            if msgpack.loads(msgpack.dumps([1, 2, 3], use_bin_type=False), use_list=True) is None:
+                raise ImportError
+        else:
+            if msgpack.loads(msgpack.dumps([1, 2, 3]), use_list=True) is None:
+                raise ImportError
+    except ImportError:
+        # Fall back to msgpack_pure
+        try:
+            import msgpack_pure as msgpack  # pylint: disable=import-error
+        except ImportError:
+            return
+    return msgpack
+
+
+# To use a custom module or version getter for the depenency, map them here.
+DEPS_MAP = {
+    'msgpack': Getters(msgpack_module_getter, None),
+    'gnupg': Getters(None, None),
+}
+
+
+class Dependencies(object):
+
+    def __init__(self, deps_map=None):
+        if deps_map is None:
+            self.deps_map = DEPS_MAP
+        else:
+            self.deps_map = deps_map
+
+    def __getattr__(self, val):
+        if val not in self.deps_map:
+            raise DependencyNotRegistered('Unknown dependency: {}'.format(val))
+        module_getter, version_getter = self.deps_map[val]
+        return Dependency(
+            val,
+            module_getter or default_module_getter,
+            version_getter or default_version_getter,
+        )
+
+
+depends = Dependencies()
```

Comment:
there's already a [`depends`](https://github.com/saltstack/salt/blob/neon/salt/utils/decorators/__init__.py#L33-L217) decorator, `dependencies`? `requirements`?

```
from salt.utils.dependencies import requirements
if requirements.gnupg >= '1.3.1':
   ...
```
doesn't look that bad...

Reply:
We also have `salt.versions` I could see putting this there too.

Reply:
That would also make sense, and I wouldn't think it would cause the same speed penalty as importing it in `salt.utils`, ie the `from salt.utils.dependencies import depends` there

Reply:
I moved this from salt.utils.depends to salt.version.reqs